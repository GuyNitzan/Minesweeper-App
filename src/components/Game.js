import React from 'react'
import Board from './Board'
import Checkbox from '@material-ui/core/Checkbox';
import withStyles from "@material-ui/core/styles/withStyles";
import green from "@material-ui/core/colors/green";
import Button from '@material-ui/core/Button';

export default class Game extends React.Component {
    state = {
        height: DefaultState.height,
        width: DefaultState.width,
        mines: DefaultState.mines,
        errorMessage: ErrorMessage.noError,
        superman: false
    };

    render () {
        const { height, width, mines, errorMessage, superman } = this.state;
        return (
        <div>
            <div>
                <div>
                    Superman<GreenCheckbox id="superman"/>
                </div>
                <div>
                    width: <input id="width" type="number" min="1" max="300" required defaultValue={DefaultState.width}/>
                    height: <input id="height" type="number" min="1" max="300" required defaultValue={DefaultState.height}/>
                    mines: <input id="mines" type="number" min="0" required defaultValue={DefaultState.mines}/>
                    <ColorButton variant="contained" onClick={this.startGame}>
                        New Game
                    </ColorButton>
                </div>
                <div className={"error"}>
                    {errorMessage}
                </div>
            </div>
            <br/>
            <div className="board">
                <Board height={height} width={width} mines={mines} errorMessage={errorMessage} superman={superman} />
            </div>
        </div>
        )
    }

    // set the relevant error message for the user input (if exists) in the state
    setErrorMessage(height, width, mines) {
        if (height.length === 0 || width.length === 0 || mines.length === 0) {
            this.setState({errorMessage: ErrorMessage.emptyField});
        } else if ((height.length>0 && height<1) || (width.length>0 && width<1)) {
            this.setState({errorMessage: ErrorMessage.negativeValue});
        } else if ((height.length>0 && height>300) || (width.length>0 && width>300)) {
            this.setState({errorMessage: ErrorMessage.maxSize});
        } else if (mines.length>0 && mines<0) {
            this.setState({errorMessage: ErrorMessage.minesMin});
        } else if (mines.length>0 && mines>height*width) {
            this.setState({errorMessage: ErrorMessage.minesMax});
        } else {
            this.setState({errorMessage: ErrorMessage.noError});
        }
    }

    // will be called upon starting a new game
    startGame = () => {
        const height = document.getElementById("height").value;
        const width = document.getElementById("width").value;
        const mines = document.getElementById("mines").value;
        this.setErrorMessage(height, width, mines);
        this.setState({
            height: height,
            width: width,
            mines: mines,
            superman: document.getElementById("superman").checked
        });
    }
}

export const DefaultState = Object.freeze({
    height: 8,
    width:8,
    mines: 10
});

export const ErrorMessage = Object.freeze({
    negativeValue: "Board height and width must have non-negative values",
    maxSize: "Board height and width cannot be bigger than 300",
    minesMax: "Mines amount cannot be bigger than height * width",
    minesMin: "Mines amount cannot be a negative number",
    emptyField: "All fields must be non-empty",
    noError: ""
});

export const GreenCheckbox = withStyles({
    root: {
        color: green[400],
        '&$checked': {
            color: green[500],
        },
    },
    checked: {},
})(props => <Checkbox color="default" {...props} />);

const ColorButton = withStyles(theme => ({
    root: {
        color: theme.palette.getContrastText(green[900]),
        backgroundColor: green[800],
        '&:hover': {
            backgroundColor: green[600],
        },
    },
}))(Button);