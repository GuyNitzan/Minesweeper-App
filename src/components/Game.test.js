import Game, {DefaultState, ErrorMessage} from './Game';
import React from 'react';
import renderer from 'react-test-renderer';

const gameTestRenderer = renderer.create(
    <Game />
);
const gameComponent = gameTestRenderer.getInstance();

test('Game component Snapshot test', () => {
    let data = gameTestRenderer.toJSON();
    expect(data).toMatchSnapshot();
});

test('Game component state error message test', () => {
    gameComponent.setErrorMessage(gameComponent.state.height, gameComponent.state.width, gameComponent.state.mines);
    expect(gameComponent.state.errorMessage).toBe(ErrorMessage.noError);
});