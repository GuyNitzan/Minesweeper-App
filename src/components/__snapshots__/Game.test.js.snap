// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game component Snapshot test 1`] = `
<div>
  <div>
    <div>
      Superman
      <span
        aria-disabled={false}
        className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-9 MuiCheckbox-root Component-root-1"
        onBlur={[Function]}
        onClick={undefined}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role={undefined}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <input
            autoFocus={undefined}
            checked={undefined}
            className="PrivateSwitchBase-input-12"
            data-indeterminate={false}
            defaultChecked={undefined}
            disabled={undefined}
            id="superman"
            name={undefined}
            onChange={[Function]}
            readOnly={undefined}
            required={undefined}
            tabIndex={undefined}
            type="checkbox"
            value={undefined}
          />
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root"
            color={undefined}
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
      </span>
    </div>
    <div>
      width: 
      <input
        defaultValue={8}
        id="width"
        max="300"
        min="1"
        required={true}
        type="number"
      />
      height: 
      <input
        defaultValue={8}
        id="height"
        max="300"
        min="1"
        required={true}
        type="number"
      />
      mines: 
      <input
        defaultValue={10}
        id="mines"
        min="0"
        required={true}
        type="number"
      />
      <button
        className="MuiButtonBase-root MuiButton-root WithStyles(ForwardRef(Button))-root-34 MuiButton-contained"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          New Game
        </span>
      </button>
    </div>
    <div
      className="error"
    >
      
    </div>
  </div>
  <br />
  <div
    className="board"
  >
    <div>
      <div>
        <div>
          Remaining Flags: 
          10
        </div>
        <br />
        <h1
          className="loseColor"
        >
           
        </h1>
      </div>
      <br />
      <div>
        <table>
          <tbody>
            <tr>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
            </tr>
            <tr>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
            </tr>
            <tr>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
            </tr>
            <tr>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
            </tr>
            <tr>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
            </tr>
            <tr>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
            </tr>
            <tr>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
            </tr>
            <tr>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
              <td
                className="hidden"
              >
                <div
                  className="cell"
                  onClick={[Function]}
                />
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
